test/upgradeableProxy/Dispatcher.upgrade.t.sol

```diff
@@ -31,48 +47,81 @@ abstract contract UpgradeTestUtils {
         bytes memory initData = abi.encodeWithSignature("initialize(string)", portPrefix);
         UUPSUpgradeable(dispatcherProxy).upgradeToAndCall(address(newDispatcherImplementation), initData);
     }
+
+    function deployDispatcherRC4ProxyAndImpl(string memory initPortPrefix, ILightClient initLightClient)
+        public
+        returns (IDispatcher proxy)
+    {
+        DispatcherRc4 dispatcherImplementation = new DispatcherRc4();
+        bytes memory initData =
+            abi.encodeWithSelector(DispatcherRc4.initialize.selector, initPortPrefix, initLightClient);
+        proxy = IDispatcher(address(new ERC1967Proxy(address(dispatcherImplementation), initData)));
+    }
 }
 
 contract ChannelHandShakeUpgradeUtil is ChannelHandshakeUtils {
     uint32 nextSequenceSendSlot = 153;
     uint32 sendPacketCommitmentSlot = 156;
+    uint32 ackPacketCommitmentSlot = 158;
     uint32 nextSequenceAckSlot = 155;
     uint32 nextSequenceRecvSlot = 154;
     IbcPacket[3] packets;
     string payload = "msgPayload";
     bytes packet = abi.encodePacked(payload);
     uint64 timeoutTimestamp = 1000;
 
-    function doChannelHandshake() public {
+    // Conduct 4-step channel handshake between local and remote end
+    // Have to pass in receivingDapp because it's hard to parse out the addresss from the portId
+    function doChannelHandshake(LocalEnd memory le, ChannelEnd memory re, IbcChannelReceiver receivingDapp) public {
         // same setup as that run in test_connectChannel_ok
         ChannelHandshakeSetting[8] memory settings = createSettings2(true);
 
+        LocalEnd memory remote =
+            LocalEnd(receivingDapp, re.portId, re.channelId, le.connectionHops, le.versionCall, le.versionExpected);
+
+        ChannelEnd memory local = ChannelEnd(le.portId, le.channelId, le.versionCall);
         string[2] memory versions = ["1.0", "2.0"];
         for (uint256 i = 0; i < settings.length; i++) {
             for (uint256 j = 0; j < versions.length; j++) {
-                LocalEnd memory le = _local;
-                ChannelEnd memory re = _remote;
                 le.versionCall = versions[j];
                 le.versionExpected = versions[j];
                 re.version = versions[j];
+                remote.versionCall = versions[j];
+                remote.versionExpected = versions[j];
+                local.version = versions[j];
                 channelOpenInit(le, re, settings[i], true);
-                channelOpenTry(le, re, settings[i], true);
+                channelOpenTry(remote, local, settings[i], true);
```

Comment:
did remote and local get reversed?

Reply:
+1

Reply:
Shouldn't they be reversed on each call, since they will be on separate sides (chanOpen init and ack happens on src chain, while chanOpen try and confirm happens on dest)? We do a similar thing of reversing them [here](https://github.com/open-ibc/vibc-core-smart-contracts/blob/25e9a3d911ee8c45218021aee2c230abb03f335b/test/VirtualChain.sol#L139-L141), so that's what I was following

Reply:
I think people including me get confused seeing the code changes. If this is the right way, then how previous "wrong" params pass the tests? If both works, then we have a bigger problem to fix

[Previous implementation](https://github.com/open-ibc/vibc-core-smart-contracts/blob/25e9a3d911ee8c45218021aee2c230abb03f335b/test/VirtualChain.sol#L139-L141) uses symmetrical params and avoid dup code. 
Tests here seems convoluted:
- `LocalEnd` already has a `IbcChannelReceiver` field.  
- Why local and remote are not the same type?`LocalEnd memory le, ChannelEnd memory re` 
- Why do multi channel handshakes in one call?
- channel version is part of `ChannelSetting`

Reply:
My original intention was to change these tests to better model the local and remote ends as seperate chains, but I see how this is very confusing  (and perhaps an incorrect test) given I'm mixing this implementation with our unitests which don't use the virtual chain.

I think the resolution here is to refactor this suite to use virtual chain, I will push these changes shortly

> If this is the right way, then how previous "wrong" params pass the tests?
The previous version assumed the local and remote ends were both on the same chain. This new version assumes that they are on different chains. It makes sense both of these two tests pass because the only access control is around proof verification, and we're using dummy clients for this test

Reply:
acutally, I think re-facorting to use Virtual chain will be slightly more complex than I originally thought. Seeing as we have a few different PRs we still need to merge, I will save this as a later task once we get everything we need sorted

Reply:
so for now, I think we can just stick to not switching local and remote for now

Reply:
resolving this for now to merge, added as an isuse on the kanban to resolve later a

Reply:
Yea, previous params were wrong then or rather testing undefined behavior. Channel handshake 4 steps should be alternating between A and B.